src/Aspire.Cli/Commands/McpStartCommand.cs

```diff
@@ -63,137 +65,356 @@ protected override async Task<int> ExecuteAsync(ParseResult parseResult, Cancell
 
-            // Get the appropriate connection using the new selection logic
-            var connection = GetSelectedConnection();
-            if (connection == null)
+            // Convert McpClientTool to Tool
+            var tools = clientTools.Select(t => t.ProtocolTool).ToList();
+
+            // Subscribe to tool list changes from the AppHost
+            await SubscribeToToolListChangesAsync(connection, cancellationToken);
+
+            if (_logger.IsEnabled(LogLevel.Debug))
             {
-                _logger.LogWarning("No Aspire AppHost is currently running");
-                throw new McpProtocolException(
-                    "No Aspire AppHost is currently running. " +
-                    "To use Aspire MCP tools, you must first start an Aspire application by running 'aspire run' in your AppHost project directory. " +
-                    "Once the application is running, the MCP tools will be able to connect to the dashboard and execute commands.",
-                    McpErrorCode.InternalError);
+                _logger.LogDebug("Fetched {ToolCount} tools from AppHost: {ToolNames}", tools.Count, string.Join(", ", tools.Select(t => t.Name)));
             }
 
-            if (connection.McpInfo == null)
+            return tools;
+        }
+        catch (Exception ex)
+        {
+            _logger.LogWarning(ex, "Failed to fetch tools from AppHost: {AppHostPath}", currentAppHostPath);
+            return null;
+        }
+    }
+
+    /// <summary>
+    /// Subscribes to tool list changes from the AppHost and forwards them to clients.
+    /// </summary>
+    private async Task SubscribeToToolListChangesAsync(AppHostAuxiliaryBackchannel connection, CancellationToken cancellationToken)
+    {
+        // Dispose previous resources if any
+        if (_toolListChangedHandler is not null)
+        {
+            await _toolListChangedHandler.DisposeAsync();
+            _toolListChangedHandler = null;
+        }
+        if (_notificationClient is not null)
+        {
+            await _notificationClient.DisposeAsync();
+            _notificationClient = null;
+        }
+
+        if (connection.McpInfo is null)
+        {
+            return;
+        }
+
+        try
+        {
+            // Create a persistent MCP client for receiving notifications
+            _notificationClient = await CreateMcpClientAsync(connection, cancellationToken);
+
+            // Register handler for tool list changes
+            _toolListChangedHandler = _notificationClient.RegisterNotificationHandler(
+                NotificationMethods.ToolListChangedNotification,
+                async (notification, ct) =>
+                {
+                    _logger.LogDebug("Received tool list changed notification from AppHost");
+
+                    // Forward the notification to our clients
+                    if (_mcpServer is not null)
```

Comment:
There are a lot of places that is using state related to the selected app host. And the app host could change while they are running.

Is it ok to continue on (this looks like it is the current behavior) or should the MCP server fail in-progress API calls when the app host changes?

In-progress calls could be cancelled when the app host changed by having composite cancellation tokens inside the API calls and changing the app host cancels the token the composite token is created from.